lib/princeton_marc.rb

```diff
@@ -149,6 +149,51 @@ def process_names record
   names.uniq
 end
 
+##
+# Get hash of authors grouped by role
+# @return [Hash]
+def process_author_roles record
+  author_roles = {
+    'TRL' => 'translators',
+    'EDT' => 'editors',
+    'COM' => 'compilers'
+  }
+  names = {}
+  names['secondary_authors'] = []
+  names['translators'] = []
+  names['editors'] = []
+  names['compilers'] = []
+
+  Traject::MarcExtractor.cached('100a:110a:111a:700a:710a:711a').collect_matching_lines(record) do |field, spec, extractor|
+    name = extractor.collect_subfields(field, spec).first
+    unless name.nil?
+      name = Traject::Macros::Marc21.trim_punctuation(name)
+
+      # If name is from 1xx field, it is the primary author.
+      if /1../.match(field.tag)
+        names['primary_author'] = name
+      else
+        field.subfields.each do |s_field|
+
+          # If field has relator code (subfield 4).
+          if s_field.code == '4'
```

Comment:
If you are on a campus IP address, you can access it. It's authenticated by IP instead of by login. But I could put it into Google Drive as well.

They are much more precise in many cases than the MARC relators. There are relator terms for contributors of works, expressions, manifestations, and items.